### PR TITLE
[WIP] map matrix

### PR DESCRIPTION
**Pull request by gouarin**

### CODE REVIEW

File: include/xtensor-sparse/xlil.hpp [hunk 1]
```diff
@@ -41,8 +59,8 @@ namespace xt
     private:
 
         using index_type = size_type;
-        using list_index_type = std::vector<index_type>;
-        using container_type = std::map<list_index_type, std::map<index_type, value_type>>;
+        using list_index_type = svector<size_type>;
+        using container_type = std::map<list_index_type, value_type>;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xlil.hpp]: I think that we can remove this definition of `index_type` and rename `list_index_type` into `index_type`

File: include/xtensor-sparse/xlil.hpp [hunk 2]
```diff
@@ -25,13 +27,29 @@ namespace xt
         using difference_type = std::ptrdiff_t;
 
         using shape_type = svector<size_type>;
+        using strides_type = svector<size_type>;
+        using backstrides_type = svector<size_type>;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xlil.hpp]: I don't think it makes sense to have strides and backstrides in sparse tensors.

File: include/xtensor-sparse/xlil.hpp [hunk 3]
```diff
@@ -52,36 +70,98 @@ namespace xt
...
         reference access_impl(Args... args);
 
-        template<class... Args>
-        std::pair<list_index_type, index_type> build_keys(Args... args) const;
+        template <class S = shape_type>
+        void reshape_impl(S&& shape, std::false_type);
 
         shape_type m_shape;
+        shape_type m_strides;
+        shape_type m_backstrides;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xlil.hpp]: I don't think it makes sense to have strides and backstrides in sparse tensors.

File: include/xtensor-sparse/xlil.hpp [hunk 4]
```diff
@@ -52,36 +70,98 @@ namespace xt
...
+    template <class S>
+    inline void xlil_container<T>::resize(S&& shape, bool force)
+    {
+        std::size_t dim = shape.size();
+        if (m_shape.size() != dim || !std::equal(std::begin(shape), std::end(shape), std::begin(m_shape)) || force)
+        {
+            m_shape = xtl::forward_sequence<shape_type, S>(shape);
+            resize_container(m_strides, dim);
+            resize_container(m_backstrides, dim);
+            compute_strides<layout_type::row_major>(m_shape, XTENSOR_DEFAULT_LAYOUT, m_strides, m_backstrides);
```
Comment (JohanMabille) [inline include/xtensor-sparse/xlil.hpp]: This should be replaced with an implementation specific to sparse containers.

File: include/xtensor-sparse/xlil.hpp [hunk 5]
```diff
@@ -52,36 +70,98 @@ namespace xt
...
+    inline void xlil_container<T>::reshape_impl(S&& shape, std::false_type /* is unsigned */)
+    {
+        if (compute_size(shape) != this->size())
+        {
+            XTENSOR_THROW(std::runtime_error, "Cannot reshape with incorrect number of elements. Do you mean to resize?");
+        }
+        m_shape = xtl::forward_sequence<shape_type, S>(shape);
+        resize_container(m_strides, m_shape.size());
+        resize_container(m_backstrides, m_shape.size());
+        compute_strides<layout_type::row_major>(m_shape, XTENSOR_DEFAULT_LAYOUT, m_strides, m_backstrides);
```
Comment (JohanMabille) [inline include/xtensor-sparse/xlil.hpp]: Same here

**Review by JohanMabille — CHANGES_REQUESTED** [review]

File: include/xtensor-sparse/xlil.hpp (hunk 2 above)
Comment (gouarin) [reply include/xtensor-sparse/xlil.hpp]: I let `strides` and `backstrides` to compute the new keys easily when we reshape. I will see if we can remove them.

File: include/xtensor-sparse/xlil.hpp (hunk 4 above)
Comment (gouarin) [reply include/xtensor-sparse/xlil.hpp]: you're right.

**Comment by gouarin** [comment]
@JohanMabille can you review ?

File: CMakeLists.txt [hunk 6]
```diff
@@ -40,7 +40,7 @@ endif()
 # =====
 
 set(XTENSOR_SPARSE_HEADERS
-    ${XTENSOR_SPARSE_INCLUDE_DIR}/xtensor-sparse/xlil.hpp
+    ${XTENSOR_SPARSE_INCLUDE_DIR}/xtensor-sparse/xmap.hpp
```
Comment (JohanMabille) [inline CMakeLists.txt]: Is that possible to name the file `xmap_container`? `xmap` sounds like a specific implementation of a standard map.

File: include/xtensor-sparse/xmap.hpp [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        using const_reference = const T&;
+        using reference = T&;
+        using pointer = T*;
+        using const_pointer = const T*;
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+
+        using shape_type = svector<size_type>;
+        using strides_type = svector<size_type>;
+        using backstrides_type = svector<size_type>;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap.hpp]: `strides_type` and `backstrides_type` should not be in the public API

File: include/xtensor-sparse/xmap.hpp [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+        using inner_shape_type = svector<size_type>;
+        using inner_strides_type = svector<size_type>;
+        using inner_backstrides_type = svector<size_type>;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap.hpp]: `inner_strides_type` and `inner_backstrides_type` can be totally removed.

File: include/xtensor-sparse/xmap.hpp [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+        template<class... Args>
+        reference operator()(Args... args);
+
+    private:
+
+        using index_type = svector<size_type>;
+        using container_type = std::map<index_type, value_type>;
+
+        // FIXME: set constexpr
+        static const value_type ZERO;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap.hpp]: Do you have any issue when setting this as `constexpr`?

File: include/xtensor-sparse/xmap.hpp [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+        {
+            XTENSOR_THROW(std::runtime_error, "Negative axis size cannot be inferred. Shape mismatch.");
+        }
+        std::decay_t<S> shape = _shape;
+        value_type accumulator = 1;
+        std::size_t neg_idx = 0;
+        std::size_t i = 0;
+        for(auto it = shape.begin(); it != shape.end(); ++it, i++)
+        {
+            auto&& dim = *it;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap.hpp]: I think that `auto` is enough here.

File: include/xtensor-sparse/xmap.hpp [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+    {
+        // TODO: check if all args have the good type
+        index_type key{static_cast<size_type>(args)...};
+
+        auto it = m_data.find(key);
+        if (it == m_data.end())
+        {
+            m_data[key] = ZERO;
+        }
+            return m_data[key];
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap.hpp]: Formatting: indent

File: include/xtensor-sparse/xmap.hpp [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+            for(std::size_t i = 0; i < shape.size(); ++i)
+            {
+                accumulator *= static_cast<size_type>(shape[i]);
+            }
+            
+            for(std::size_t i = 0; i < shape.size(); ++i)
+            {
+                accumulator /= static_cast<size_type>(shape[i]);
+                new_key[i] = index/accumulator;
+                index -= new_key[i]*accumulator;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap.hpp]: What about storing the corresponding strides, and reuse element_offset and unravel_index?

**Review by JohanMabille — CHANGES_REQUESTED** [review]

File: include/xtensor-sparse/xmap.hpp (hunk 7 above)
Comment (gouarin) [reply include/xtensor-sparse/xmap.hpp]: we don't need it anymore. I forget to remove these types.

File: include/xtensor-sparse/xmap.hpp (hunk 10 above)
Comment (gouarin) [reply include/xtensor-sparse/xmap.hpp]: this line comes from the implementation of `reshape_impl` for an `xstrided_container`.

File: include/xtensor-sparse/xmap.hpp (hunk 12 above)
Comment (gouarin) [reply include/xtensor-sparse/xmap.hpp]: It's a good idea. I didn't know these functions!

File: include/xtensor-sparse/xmap_container.hpp [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+        using inner_shape_type = svector<size_type>;
+
+        xmap_container(const shape_type& shape);
+
+        size_type size() const noexcept;
+
+        const shape_type& shape() const;
+        size_type dimension() const;
+        const strides_type& strides() const;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap_container.hpp]: I still think we should not expose the strides. They are just a convenience for computing new keys when reshaping.

File: include/xtensor-sparse/xmap_container.hpp [hunk 14]
```diff
@@ -0,0 +1,278 @@
+#ifndef XTENSOR_SPARSE_LIL_HPP
+#define XTENSOR_SPARSE_LIL_HPP
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap_container.hpp]: `XTENSOR_SPARSE_LIL_HPP` -> `XSPARSE_MAP_CONTAINER_HPP` (less cumbersome than `XTENSOR_SPARSE_MAP_CONTAINER_HPP`

File: include/xtensor-sparse/xmap_container.hpp [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+        using value_type = T;
+        using const_reference = const T&;
+        using reference = T&;
+        using pointer = T*;
+        using const_pointer = const T*;
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+
+        using shape_type = svector<size_type>;
+        using strides_type = svector<size_type>;
```
Comment (JohanMabille) [inline include/xtensor-sparse/xmap_container.hpp]: This should move to the private section

**Review by JohanMabille — APPROVED** [review]

**Pull request merged**